src/tools/U8U16Test/main.cpp

```diff
@@ -0,0 +1,429 @@
+    ULONG written{};
+    ULONG total{};
+    NTSTATUS status{};
+    GetDuration();
+    std::unique_ptr<wchar_t[]> u16Buffer{ std::make_unique<wchar_t[]>(u8Str.length()) };
+    duration += GetDuration();
+
+    for (size_t i{}; i < endLoop; i += u8CharLen)
+    {
+        const std::string_view sv{ &u8Str.at(i), u16ChunkLen * u8CharLen };
+        GetDuration();
+        status = p_RtlUTF8ToUnicodeN(u16Buffer.get(), static_cast<ULONG>(sv.length() * sizeof(wchar_t)), &written, sv.data(), static_cast<ULONG>(sv.length()));
+        duration += GetDuration();
+        total += written;
+    }
+
+    const wchar_t randElem16 = u16Buffer[RandomIndex(static_cast<ptrdiff_t>(u16ChunkLen))];
+    u16Buffer.reset();
+    std::cout << " ignore me " << static_cast<int>(randElem16)
+              << "\n NTSTATUS " << status << "\n length " << (total / sizeof(wchar_t)) << "\n elapsed " << duration << std::endl;
+}
+
+void u8u16_Chunks(std::string_view u8Str, size_t u8CharLen, size_t u16ChunkLen)
+{
+    PrintHeader(__func__);
+    const size_t endLoop{ u8Str.length() / u16ChunkLen };
+    double duration{};
+    size_t length{};
+    HRESULT hRes{};
+    std::wstring u16Str{};
+
+    for (size_t i{}; i < endLoop; i += u8CharLen)
+    {
+        const std::string_view sv{ &u8Str.at(i), u16ChunkLen * u8CharLen };
+        GetDuration();
+        hRes = u8u16(sv, u16Str);
+        duration += GetDuration();
+        length += u16Str.length();
+    }
+
+    const wchar_t randElem16 = u16Str.at(RandomIndex(static_cast<ptrdiff_t>(u16Str.length())));
+    std::cout << " ignore me " << static_cast<int>(randElem16)
+              << "\n HRESULT " << hRes << "\n length " << length << "\n elapsed " << duration << std::endl;
+}
+
+void u8u16_ptr_Chunks(std::string_view u8Str, size_t u8CharLen, size_t u16ChunkLen)
+{
+    PrintHeader(__func__);
+    const size_t endLoop{ u8Str.length() / u16ChunkLen };
+    double duration{};
+    size_t length{};
+    HRESULT hRes{};
+    std::wstring u16Str{};
+
+    for (size_t i{}; i < endLoop; i += u8CharLen)
+    {
+        const std::string_view sv{ &u8Str.at(i), u16ChunkLen * u8CharLen };
+        GetDuration();
+        hRes = u8u16_ptr(sv, u16Str);
+        duration += GetDuration();
+        length += u16Str.length();
+    }
+
+    const wchar_t randElem16 = u16Str.at(RandomIndex(static_cast<ptrdiff_t>(u16Str.length())));
+    std::cout << " ignore me " << static_cast<int>(randElem16)
+              << "\n HRESULT " << hRes << "\n length " << length << "\n elapsed " << duration << std::endl;
+}
+
+int main()
+{
+    // UTF-16 string length
+    constexpr const size_t u16Length{ 100000000u }; // 100,000 code points
+    //constexpr const size_t u16Length{ 10000000u }; // 10,000 code points
+
+    // chunk length in code points
+    constexpr const size_t chunkLen = 10u;
+
+    // UTF-16 character to be used
+    //const std::wstring testU16(u16Length, static_cast<wchar_t>(0x007E)); // TILDE (1 Byte in UTF-8)
```

Comment:
nit: I agree that a broader diversity of test input might be helpful here for benchmarking

Reply:
I was on a business trip the past few days and didn't have time to implement tests for natural languages. I'll catch up over weekend.